add duration when sending cr

paolochiodi opened the pull request.

paolochiodi committed:
add duration when sending cr

codefromthecrypt commented on test/zipkin-simple.test.js:
```diff
@@ -134,7 +134,7 @@ describe("Zipkin client", function () {
 				spanId: "test spanId",
 				parentSpanId: "test parent_id",
-				timestamp: "test timestamp",
+				timestamp: new Date().getTime(),
```
is this in microseconds? if it is milliseconds, multiply by 1000 (just to set a good example)

paolochiodi replied:
you're right, this is in milliseconds. 
This doesn't actually effect the test because we are using it only for the duration, but I'll add \* 1000 for clarity

codefromthecrypt commented on zipkin-simple.js:
```diff
@@ -127,6 +127,10 @@ function sendTrace (trace, data) {
 		})
+
+		if (data.annotations[i] === "cr") {
+			body.duration = time - trace.timestamp
```
did you want to add body.timestamp, too, or is that done somewhere else?

paolochiodi replied:
the timestamp of the start time is already added, the timestamp of the end is added a the timestamp of the cr annotation (on line 127).

Is this the correct way for zipkin or should I also update the span timestamp?

codefromthecrypt replied:
so, if the goal is to add duration when you "own" the span, the same applies to timestamp.

otherwise, the same span id will store two timestamps.. one for "cs" and one for "sr". make sense?

paolochiodi committed:
make method names more explicit

paolochiodi committed:
set better example in tests using microseconds

codefromthecrypt commented:
hey I was just thinking. the only special case about timestamp+duration is really the server side. This could make the logic simpler. ex blacklist instead of whitelist.

so something like

```
unless (logging "sr" or "ss" in an existing span) {
 log Span.timestamp and duration
}
```

food for thought

paolochiodi commented:
The current implementation (as in this PR) is that:
- on the first appearance of the trace add timestamp (this should usually be on cs, or sr if the client doesn't support tracing)
- on any annotation add the timestamp to that annotation
- on cr add duration

The only thing left out is adding the duration on ss if the client doesn't support tracing. The problem is that at the moment the traceData doesn't track all the sent data, so there's no way to know if the client is tracing or not.

I'm still thinking on what's the best interface to support that case

codefromthecrypt commented:
your call, but I think this is the best logic for where things are right now.

Span.timestamp = first annotation unless "sr"
Span.duration = last annotation unless "ss" - timestamp

That contains special-casing to the only special span (the server span). Then you might put a TODO to figure out if a trace was propagated or not.

paolochiodi commented:
Ok, I have a better picture here now. 
I'm grateful for this discussion as it's highlighting a couple of weaknesses of my implementation.

The two problems here:
1- I don't have the history of a span, and it's thus difficult to understand if this is a server-only request or if any specific annotation has been made
2- I'm putting the original timestamp into the trace data that need to be handled down to other systems (and relying it on being present), however that is not compliant with the zipkin specification

Given the above, I'm afraid @adriancole solution isn't doable with current implementation as I don't have any notion of "is this the first annotation?"

I need some though/experimentation on other possible implementations

paolochiodi committed:
Change to API to accomodate correct duration reporting

- drop getData method, this is done automatically by the sendXXX
  methods
- timestamp only added on cs or sr for server only traces
- duration only added on cr or ss for server only traces
- mark a trace as server only when sending without previous trace data
- getChild still work as expected, always drop serverOnly attribute
  (assumption is that if you call getChild you are going to make a
client call)

paolochiodi committed:
update readme

paolochiodi committed:
additional README updates

codefromthecrypt commented on test/zipkin-simple.test.js:
```diff
@@ -11,20 +11,25 @@ const it = lab.it
 const expect = Code.expect
 
 const FAKE_SERVER_PORT = 9090
+const TO_MICROSENCODS = 1000
```
typo

paolochiodi replied:
good spot, thanks

codefromthecrypt commented on test/zipkin-simple.test.js:
```diff
@@ -168,39 +150,52 @@ describe("Zipkin client", function () {
 				})
 
-				Client.clientSend(traceData, {
+				Client.sendClientSend(traceData, {
```
sounds a bit wordy.. maybe startSpan?

paolochiodi replied:
I agree it's a bit wordy, but I preferred to be more descriptive here.
I added the send prefix to avoid confusion especially on what the methods were doing (i.e. clientRecv could give the idea of waiting for something to be received).

I kinda like startSpan, but I want to keep it coherent with the other methods.

Another good options could be to have startSpan, startChildSpan (which would make getChild obsolete), and finishSpan - but I can't come up with something similar for the server annotations

codefromthecrypt commented on test/zipkin-simple.test.js:
```diff
@@ -234,26 +230,26 @@ describe("Zipkin client", function () {
 				})
 
-				Client.clientRecv(traceData, {
+				Client.sendClientRecv(traceData, {
```
maybe finish? seems odd to pass the service and name explicitly here

paolochiodi replied:
For the name, see the comment above.

For the service and name explicitly here, why? Do you think I should store them in the returned traceData and reuse them?

codefromthecrypt replied:
don't mind me much I don't usually write javascript so I've no idea what idiomatic is. Just it seems odd to carry around the span name multiple times.

paolochiodi replied:
If I don't send the name on non cs, will the zipkin api accept the requests and correctly parse them?

codefromthecrypt replied:
well you shouldn't send anything to zipkin until the span is complete (so you'd have the name from "cs" but don't need it again in "cr"

codefromthecrypt commented on zipkin-simple.js:
```diff
@@ -127,34 +118,53 @@ function sendTrace (trace, data) {
 			value: data.annotations[i]
 		})
+
+		if (data.annotations[i] === "cr" || (data.annotations[i] === "ss" && trace.serverOnly)) {
```
seems right

paolochiodi commented:
I've updated the PR with a overhaul of the interface, and I think I covered all the cases for a correct timestamp and duration reporting:
- drop getData method, this is done automatically by the sendXXX
  methods
- timestamp only added on cs or sr for server only traces
- duration only added on cr or ss for server only traces
- mark a trace as server only when sending sr without previous trace data
- getChild still work as expected, always drop serverOnly attribute
  (assumption is that if you call getChild you are going to make a
  client call)

The overall correctness of the reporting still lies on the developer, because it needs to correctly store the traceData returned by each method and pass it back to the following one - however this is one of the initial assumptions for this project: letting the developer manage and store the trace data (I think we can say the primary user for this library is someone building instrumentations for other libraries)

codefromthecrypt reviewed (commented):
time logic looks alright, just feel mixed about the function names

drusellers commented:
🎉
